### PR TITLE
Make Bootstrap tool use Helix Cluster Manager for validation

### PR DESCRIPTION
**Pull request by pnarayanan**
Instantiate a Helix cluster manager within the bootstrap tool to listen
on the changes that it makes itself. This cluster manager's state will
be checked during validation. If there are other errors that can cause
errors such as Out of Memory in the Helix client, then those can be
immediately detected because the tool itself will be affected.

### CODE REVIEW

**Comment by pnarayanan** [comment]
@jsjtzyy fixed test failures and rebased with master. Good to review when you get a chance. Thanks.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 1]
```diff
@@ -702,6 +711,17 @@ private void maybeAddCluster() {
...
   }
 
+  private void startClusterManager() throws IOException {
+    Properties props = new Properties();
+    props.setProperty("clustermap.host.name", "localhost");
+    props.setProperty("clustermap.cluster.name", clusterName);
+    props.setProperty("clustermap.datacenter.name", adminForDc.keySet().iterator().next());
+    props.setProperty("clustermap.dcs.zk.connect.strings", Utils.readStringFromFile(zkLayoutPath));
+    ClusterMapConfig clusterMapConfig = new ClusterMapConfig(new VerifiableProperties(props));
+    HelixClusterAgentsFactory helixClusterAgentsFactory = new HelixClusterAgentsFactory(clusterMapConfig, null, null);
```
Comment (jsjtzyy) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: just to confirm, the validating HelixClusterManager is going to talk to one zookeeper in one of the specified datacenters? Here "specified" means the selected DCs which will be updated/affected by Helix tool. And do you think it's worth setting up a second validating HelixClusterManager to simulate those in untouched DCs?

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 2]
```diff
@@ -719,7 +739,14 @@ private void validateAndClose() throws Exception {
     try {
```
Comment (jsjtzyy) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: no Exception is thrown in this method

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java [hunk 3]
```diff
@@ -433,21 +441,26 @@ AmbryReplica getReplicaForPartitionOnNode(String hostname, int port, String part
...
-        } else {
-          updateStateOfReplicas(configs);
+          sealedStateChangeCounter.incrementAndGet();
+          helixClusterManagerMetrics.instanceConfigChangeTriggerCount.inc();
         }
-        sealedStateChangeCounter.incrementAndGet();
-        helixClusterManagerMetrics.instanceConfigChangeTriggerCount.inc();
+      } catch (Throwable t) {
+        errorCount.incrementAndGet();
+        throw t;
```
Comment (jsjtzyy) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java]: In actual node (not the validating one), how the `Throwable` would be handled?

File: ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java [hunk 4]
```diff
@@ -172,7 +177,8 @@ public void testIncompleteZKHostInfo() throws Exception {
       Utils.writeJsonObjectToFile(testPartitionLayout.getPartitionLayout().toJSONObject(), partitionLayoutPath);
       try {
         HelixBootstrapUpgradeUtil.bootstrapOrUpgrade(hardwareLayoutPath, partitionLayoutPath, zkLayoutPath,
-            CLUSTER_NAME_PREFIX, dcStr, DEFAULT_MAX_PARTITIONS_PER_RESOURCE, false, false, new HelixAdminFactory());
+            CLUSTER_NAME_PREFIX, dcStr, DEFAULT_MAX_PARTITIONS_PER_RESOURCE, false, false, new HelixAdminFactory(),
+            true);
```
Comment (jsjtzyy) [inline ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java]: I think, maybe, we should also check the error count from validating HelixClusterManager. From current tests, I don't see the verification on functionality of validating HelixClusterManager.

Comment (pnarayanan) [reply ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java]: `HelixBootstrapUpgradeUtil#validateAndClose()` checks for it, are you suggesting something else?

Comment (jsjtzyy) [reply ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java]: I was suggesting getting the error count from `validatingHelixClusterManager` after running the tool. Looks like the validating piece of code in `validateAndClose()` (see follow) is not covered by this test. 
```
if (validatingHelixClusterManager != null) {
        ensureOrThrow(validatingHelixClusterManager.getErrorCount() == 0,
            "Helix cluster manager should not have encountered any errors");
      }
```

**Pull request merged**